### ISSUE TITLE
exponentially long exception messages in multi-threaded setup (ref  #37392)

### ISSUE THREAD

**Issue by slava77** [cms-sw/cmssw#37579]
I observed a problem similar to #37392 
with 
```
Exception Message:
basic_string::_M_construct null not valid
```
in 12_4_0_pre2.
While the issue goes away in a recent IB (e.g. in CMSSW_12_4_X_2022-04-14-1100), 

I have a tangential question: the exception messages in 12_4_0_pre2 are issued for every thread with increasingly long stack. In a 32-thread test the last one comes out with 
```
   [10210] While reading from source edm::HepMCProduct generatorSmeared '' SIM
   [10211] Reading branch edmHepMCProduct_generatorSmeared__SIM.
Exception Message:
basic_string::_M_construct null not valid
----- End Fatal Exception -------------------------------------------------
```
Is the framework really creating such deep stacks or is this just a feature of the printout?
I have a reference setup on cmsdev34:/build/slava77/reltest/CMSSW_12_4_0_pre2-orig/test/gensim-1210p5/11634.0_TTbar_14TeV+2021/sort
there with 8-threads the last exception message is 1703-deep

**Comment by makortel** [comment]
assign core

**Comment by makortel** [comment]
Looking at `cmsdev34:/build/slava77/reltest/CMSSW_12_4_0_pre2-orig/test/gensim-1210p5/11634.0_TTbar_14TeV+2021/sort/step2.mt8.log` I see
* 1st trace has depth 4
* 2nd trace has depth 264
* 3rd trace has depth 282
* 4th trace has depth 119
* 5th trace has depth 638
* 6th trace has depth 994
* 7th trace has depth 1347
* 8th trace has depth 1703

The traces are full of `Rethrowing an exception that happened on a different read request.`

I guess what happens is something along
* the delayed read throws an exception, one of the streams reports that
* the same exception propagates to all other delayed read requests, and those add additional context
* the additional context gets duplicated as well

**Comment by makortel** [comment]
The delayed reader throws its exception on "different read request" here
https://github.com/cms-sw/cmssw/blob/086e2e24704b05220277c52673a7bdba5e53c16c/IOPool/Input/src/RootDelayedReader.cc#L42-L49

The message gets extended with `While reading from` in
https://github.com/cms-sw/cmssw/blob/086e2e24704b05220277c52673a7bdba5e53c16c/FWCore/Framework/src/ProductResolvers.cc#L196-L204

**Comment by slava77** [comment]
Do I understand this correctly that this is just a length of a message, not an actual depth of the calls?

**Comment by makortel** [comment]
Correct, it is how the exception messages are constructed (specifically how additional context information gets added), not the depth of the actual call stack.

**Comment by slava77** [comment]
> Correct, it is how the exception messages are constructed (specifically how additional context information gets added), not the depth of the actual call stack.

OK, I guess I can close this issue as not an issue.
Thanks for checking/clarifying.

**Comment by Dr15Jones** [comment]
Should be fixed by #37591
